From 44de02bcf7ab3dfe385f2bf531285e4927e38ac1 Mon Sep 17 00:00:00 2001
From: Fabian Vogt <fabian@ritter-vogt.de>
Date: Sun, 5 Jul 2020 17:45:35 +0200
Subject: [PATCH 3/3] Move VT setup to sddm-helper

Previously, sddm set up the virtual terminal and jumped to it and after that,
sddm-helper opens it again and takes control of it. This only worked because
sddm leaked the terminal fd into sddm-helper.

The right way is to set up the virtual terminal and take control of it before
jumping to it. This is achieved by moving this responsibility to sddm-helper.

As the race is now fixed, also avoid leaking FDs in jumpToVt now.

Fixes #1078
Fixes #1083
---
 src/common/Session.cpp                         | 10 ----------
 src/common/Session.h                           |  4 ----
 src/{daemon => common}/VirtualTerminal.cpp     |  2 ++
 src/{daemon => common}/VirtualTerminal.h       |  0
 .../VirtualTerminal_FreeBSD.cpp                |  0
 src/daemon/CMakeLists.txt                      |  4 ++--
 src/daemon/Display.cpp                         | 18 +++---------------
 src/helper/CMakeLists.txt                      |  8 ++++++++
 src/helper/HelperApp.cpp                       |  6 ++++++
 src/helper/UserSession.cpp                     |  6 +++++-
 10 files changed, 26 insertions(+), 32 deletions(-)
 rename src/{daemon => common}/VirtualTerminal.cpp (99%)
 rename src/{daemon => common}/VirtualTerminal.h (100%)
 rename src/{daemon => common}/VirtualTerminal_FreeBSD.cpp (100%)

diff --git a/src/common/Session.cpp b/src/common/Session.cpp
index aa4dad3..2d49430 100644
--- a/src/common/Session.cpp
+++ b/src/common/Session.cpp
@@ -52,16 +52,6 @@ namespace SDDM {
         return m_type;
     }
 
-    int Session::vt() const
-    {
-        return m_vt;
-    }
-
-    void Session::setVt(int vt)
-    {
-        m_vt = vt;
-    }
-
     QString Session::xdgSessionType() const
     {
         return m_xdgSessionType;
diff --git a/src/common/Session.h b/src/common/Session.h
index d285a3f..9fd86cd 100644
--- a/src/common/Session.h
+++ b/src/common/Session.h
@@ -42,9 +42,6 @@ namespace SDDM {
 
         Type type() const;
 
-        int vt() const;
-        void setVt(int vt);
-
         QString xdgSessionType() const;
 
         QDir directory() const;
@@ -69,7 +66,6 @@ namespace SDDM {
     private:
         bool m_valid;
         Type m_type;
-        int m_vt;
         QDir m_dir;
         QString m_name;
         QString m_fileName;
diff --git a/src/daemon/VirtualTerminal.cpp b/src/common/VirtualTerminal.cpp
similarity index 99%
rename from src/daemon/VirtualTerminal.cpp
rename to src/common/VirtualTerminal.cpp
index 196e387..4607c50 100644
--- a/src/daemon/VirtualTerminal.cpp
+++ b/src/common/VirtualTerminal.cpp
@@ -187,6 +187,8 @@ out:
                 qWarning("Couldn't finalize jump to VT %d: %s", vt, strerror(errno));
 
             close(activeVtFd);
+            if (fd != activeVtFd)
+                close(fd);
         }
     }
 }
diff --git a/src/daemon/VirtualTerminal.h b/src/common/VirtualTerminal.h
similarity index 100%
rename from src/daemon/VirtualTerminal.h
rename to src/common/VirtualTerminal.h
diff --git a/src/daemon/VirtualTerminal_FreeBSD.cpp b/src/common/VirtualTerminal_FreeBSD.cpp
similarity index 100%
rename from src/daemon/VirtualTerminal_FreeBSD.cpp
rename to src/common/VirtualTerminal_FreeBSD.cpp
diff --git a/src/daemon/CMakeLists.txt b/src/daemon/CMakeLists.txt
index a175176..9145607 100644
--- a/src/daemon/CMakeLists.txt
+++ b/src/daemon/CMakeLists.txt
@@ -36,9 +36,9 @@ set(DAEMON_SOURCES
 # Different implementations of the VT switching code
 # (where the FreeBSD version does nothing).
 if(${CMAKE_SYSTEM} MATCHES "FreeBSD")
-    list(APPEND DAEMON_SOURCES VirtualTerminal_FreeBSD.cpp)
+    list(APPEND DAEMON_SOURCES ${CMAKE_SOURCE_DIR}/src/common/VirtualTerminal_FreeBSD.cpp)
 else()
-    list(APPEND DAEMON_SOURCES VirtualTerminal.cpp)
+    list(APPEND DAEMON_SOURCES ${CMAKE_SOURCE_DIR}/src/common/VirtualTerminal.cpp)
 endif()
 
 qt5_add_dbus_adaptor(DAEMON_SOURCES "${CMAKE_SOURCE_DIR}/data/interfaces/org.freedesktop.DisplayManager.xml"          "DisplayManager.h" SDDM::DisplayManager)
diff --git a/src/daemon/Display.cpp b/src/daemon/Display.cpp
index 6cd8fd6..6a2c9a6 100644
--- a/src/daemon/Display.cpp
+++ b/src/daemon/Display.cpp
@@ -30,7 +30,6 @@
 #include "Greeter.h"
 #include "Utils.h"
 #include "SignalHandler.h"
-#include "VirtualTerminal.h"
 
 #include <QDebug>
 #include <QFile>
@@ -310,14 +309,10 @@ namespace SDDM {
 
         QProcessEnvironment env;
 
-        // create new VT for Wayland sessions otherwise use greeter vt
         if (seat()->name() == QLatin1String("seat0")) {
-            int vt = terminalId();
-            if (session.xdgSessionType() == QLatin1String("wayland"))
-                vt = VirtualTerminal::setUpNewVt();
-            m_lastSession.setVt(vt);
-            env.insert(QStringLiteral("XDG_VTNR"), QString::number(vt));
-	}
+            // Use the greeter VT, for wayland sessions the helper overwrites this
+            env.insert(QStringLiteral("XDG_VTNR"), QString::number(terminalId()));
+        }
 
         env.insert(QStringLiteral("PATH"), mainConfig.Users.DefaultPath.get());
         if (session.xdgSessionType() == QLatin1String("x11"))
@@ -363,13 +358,6 @@ namespace SDDM {
                 stateConfig.Last.Session.setDefault();
             stateConfig.save();
 
-            // switch to the new VT for Wayland sessions
-            if (seat()->name() == QLatin1String("seat0")) {
-                if (m_lastSession.xdgSessionType() == QLatin1String("wayland"))
-                    // set vt_auto to false, so handle the vt switch yourself (VT_PROCESS)
-                    VirtualTerminal::jumpToVt(m_lastSession.vt(), false);
-            }
-
             if (m_socket)
                 emit loginSucceeded(m_socket);
         } else if (m_socket) {
diff --git a/src/helper/CMakeLists.txt b/src/helper/CMakeLists.txt
index 3dce2ab..81b939b 100644
--- a/src/helper/CMakeLists.txt
+++ b/src/helper/CMakeLists.txt
@@ -17,6 +17,14 @@ set(HELPER_SOURCES
     UserSession.cpp
 )
 
+# Different implementations of the VT switching code
+# (where the FreeBSD version does nothing).
+if(${CMAKE_SYSTEM} MATCHES "FreeBSD")
+    list(APPEND HELPER_SOURCES ${CMAKE_SOURCE_DIR}/src/common/VirtualTerminal_FreeBSD.cpp)
+else()
+    list(APPEND HELPER_SOURCES ${CMAKE_SOURCE_DIR}/src/common/VirtualTerminal.cpp)
+endif()
+
 if(PAM_FOUND)
     set(HELPER_SOURCES
         ${HELPER_SOURCES}
diff --git a/src/helper/HelperApp.cpp b/src/helper/HelperApp.cpp
index 13ccdfa..33853f5 100644
--- a/src/helper/HelperApp.cpp
+++ b/src/helper/HelperApp.cpp
@@ -24,6 +24,7 @@
 #include "SafeDataStream.h"
 
 #include "MessageHandler.h"
+#include "VirtualTerminal.h"
 
 #include <QtCore/QTimer>
 #include <QtCore/QFile>
@@ -148,6 +149,11 @@ namespace SDDM {
 
         if (!m_session->path().isEmpty()) {
             env.insert(m_session->processEnvironment());
+            // Allocate a new VT for the wayland session
+            if(env.value(QStringLiteral("XDG_SESSION_TYPE")) == QLatin1String("wayland")) {
+                int vtNumber = VirtualTerminal::setUpNewVt();
+                env.insert(QStringLiteral("XDG_VTNR"), QString::number(vtNumber));
+            }
             m_session->setProcessEnvironment(env);
 
             if (!m_backend->openSession()) {
diff --git a/src/helper/UserSession.cpp b/src/helper/UserSession.cpp
index 5e9ac6a..89d659f 100644
--- a/src/helper/UserSession.cpp
+++ b/src/helper/UserSession.cpp
@@ -23,6 +23,7 @@
 #include "UserSession.h"
 #include "HelperApp.h"
 #include "XauthUtils.h"
+#include "VirtualTerminal.h"
 
 #include <sys/types.h>
 #include <sys/ioctl.h>
@@ -82,7 +83,8 @@ namespace SDDM {
         // that it stays open without races
         if (sessionType == QLatin1String("wayland")) {
             // open VT and get the fd
-            QString ttyString = QStringLiteral("/dev/tty%1").arg(processEnvironment().value(QStringLiteral("XDG_VTNR")));
+            int vtNumber = processEnvironment().value(QStringLiteral("XDG_VTNR")).toInt();
+            QString ttyString = QStringLiteral("/dev/tty%1").arg(vtNumber);
             int vtFd = ::open(qPrintable(ttyString), O_RDWR | O_NOCTTY);
 
             // when this is true we'll take control of the tty
@@ -112,6 +114,8 @@ namespace SDDM {
                     exit(Auth::HELPER_OTHER_ERROR);
                 }
             }
+
+            VirtualTerminal::jumpToVt(vtNumber, false);
         }
 
 #ifdef Q_OS_LINUX
-- 
2.25.1

